src/index/ivf_flat.rs

```diff
@@ -0,0 +1,354 @@
+        nlist: u32,
+        metric: MetricType,
+        own_fields: bool,
+    ) -> Result<Self> {
+        unsafe {
+            let metric = metric as c_uint;
+            let mut inner = ptr::null_mut();
+            faiss_try(faiss_IndexIVFFlat_new_with_metric(
+                &mut inner,
+                quantizer.inner_ptr(),
+                d as usize,
+                nlist as usize,
+                metric,
+            ))?;
+            faiss_IndexIVFFlat_set_own_fields(inner, own_fields as i32);
+            Ok(IVFFlatIndexImpl { inner })
+        }
+    }
+
+    /// Create a new IVF flat index.
+    // The index owns the quantizer.
+    pub fn new(quantizer: flat::FlatIndex, d: u32, nlist: u32, metric: MetricType) -> Result<Self> {
+        let index = IVFFlatIndexImpl::new_helper(&quantizer, d, nlist, metric, true)?;
+        std::mem::forget(quantizer);
+
+        Ok(index)
+    }
+
+    /// Create a new IVF flat index with L2 as the metric type.
+    pub fn new_l2_by_ref(quantizer: &flat::FlatIndex, d: u32, nlist: u32) -> Result<Self> {
+        IVFFlatIndexImpl::new_by_ref(quantizer, d, nlist, MetricType::L2)
+    }
+
+    /// Create a new IVF flat index with L2 as the metric type.
+    // The index owns the quantizer.
+    pub fn new_l2(quantizer: flat::FlatIndex, d: u32, nlist: u32) -> Result<Self> {
+        IVFFlatIndexImpl::new(quantizer, d, nlist, MetricType::L2)
+    }
+
+    /// Create a new IVF flat index with IP (inner product) as the metric type.
+    pub fn new_ip_by_ref(quantizer: &flat::FlatIndex, d: u32, nlist: u32) -> Result<Self> {
+        IVFFlatIndexImpl::new_by_ref(quantizer, d, nlist, MetricType::InnerProduct)
+    }
+
+    /// Create a new IVF flat index with IP (inner product) as the metric type.
+    // The index owns the quantizer.
+    pub fn new_ip(quantizer: flat::FlatIndex, d: u32, nlist: u32) -> Result<Self> {
+        IVFFlatIndexImpl::new(quantizer, d, nlist, MetricType::InnerProduct)
+    }
+
+    /// Get number of probes at query time
+    pub fn nprobe(&self) -> u32 {
+        unsafe { faiss_IndexIVFFlat_nprobe(self.inner_ptr()) as u32 }
+    }
+
+    /// Set number of probes at query time
+    pub fn set_nprobe(&mut self, value: u32) {
+        unsafe {
+            faiss_IndexIVFFlat_set_nprobe(self.inner_ptr(), value as usize);
+        }
+    }
+
+    /// Get number of possible key values
+    pub fn nlist(&self) -> u32 {
+        unsafe { faiss_IndexIVFFlat_nlist(self.inner_ptr()) as u32 }
+    }
+
+    /// Get train type
+    pub fn train_type(&self) -> Option<TrainType> {
+        unsafe {
+            let code = faiss_IndexIVFFlat_quantizer_trains_alone(self.inner_ptr());
+            TrainType::from_code(code)
+        }
+    }
+}
+
+/**
+ * = 0: use the quantizer as index in a kmeans training
+ * = 1: just pass on the training set to the train() of the quantizer
+ * = 2: kmeans training on a flat index + add the centroids to the quantizer
+ */
+#[derive(Debug, Clone)]
```

Comment:
We can derive a few more traits on this one, as per the [C-COMMON-TRAITS](https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits) guideline.
```suggestion
#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq)]
```